drivers/flash/flash_gd32.c

```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright (c) 2022 BrainCo Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT gd_gd32_flash_controller
+
+#include <errno.h>
+#include <kernel.h>
+#include <devicetree.h>
+#include <drivers/flash.h>
+#include <soc.h>
+
+#include "flash_gd32.h"
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(flash_gd32, CONFIG_FLASH_LOG_LEVEL);
+
+struct flash_gd32_data {
+	struct k_sem mutex;
+};
+
+static struct flash_gd32_data flash_data;
+
+static const struct flash_parameters flash_gd32_parameters = {
+	.write_block_size = SOC_NV_FLASH_WRITE_BLOCK_SIZE,
+	.erase_value = 0xff,
+};
+
+static inline void flash_gd32_lock(void)
+{
+#if defined(CONFIG_SOC_SERIES_GD32F403)
+	FMC_CTL0 |= FMC_CTL0_LK;
+#if SOC_NV_FLASH_SIZE > FMC_BANK0_SIZE
+	FMC_CTL1 |= FMC_CTL1_LK;
+#endif
+#else /* !CONFIG_SOC_SERIES_GD32F403 */
+	FMC_CTL |= FMC_CTL_LK;
+#endif
+}
+
+static inline void flash_gd32_unlock(void)
+{
+#if defined(CONFIG_SOC_SERIES_GD32F403)
+	FMC_KEY0 = UNLOCK_KEY0;
+	FMC_KEY0 = UNLOCK_KEY1;
+#if SOC_NV_FLASH_SIZE > FMC_BANK0_SIZE
+	FMC_KEY1 = UNLOCK_KEY0;
+	FMC_KEY1 = UNLOCK_KEY1;
+#endif
+#else /* !CONFIG_SOC_SERIES_GD32F403 */
+	FMC_KEY = UNLOCK_KEY0;
+	FMC_KEY = UNLOCK_KEY1;
+#endif
+}
+
+static bool flash_gd32_valid_range(off_t offset, size_t len)
+{
+	if (offset > SOC_NV_FLASH_SIZE) {
+		return false;
+	}
+
+	if (len && ((offset + (len - 1)) > SOC_NV_FLASH_SIZE)) {
```

Comment:
`(offset + len) > SOC_NV_FLASH_SIZE`, for example having `SOC_NV_FLASH_SIZE == 4`, offset == 2 and len == 2 is completely OK.
Should `len == 0` return false?

Reply:
> `(offset + len) > SOC_NV_FLASH_SIZE`, for example having `SOC_NV_FLASH_SIZE == 4`, offset == 2 and len == 2 is completely OK.

done.

>Should `len == 0` return false?

I think `len == 0` should return true, there doesn't have any explicit error in here, just no nothing.